0003-configuration-management-extension/README.md

```diff
@@ -0,0 +1,140 @@
+---
+
+## Introduction
+
+The Configuration Management Extension is a [Kaa Protocol](/0001-kaa-protocol/README.md) extension.
+
+It is intended to manage endpoint configuration distribution.
+
+## Requirements and constraints
+### Problems and possible solutions
+
+1. _Multiple configurations available for client._ It is possible that there will be more than one available configurations for endpoint (e.g., endpoint was offline while new configurations were added).
+   
+   Solution:
+   - Only latest configuration should be sent to an endpoint to reduce amount of traffic.
+
+2. _The server should know if a configuration has been delivered._ 
+
+   Solutions:
+   - Send PUBLISH message from client to "/status" topic. See below sections for details.
+
+## Use cases
+
+### UC1
+Configuration delivery by request. The endpoint should be able to receive latest configuration from CMX by request.
+
+### UC2
+Configuration delivery that is initiated by the server. The endpoint should receive latest configuration when it connects to a server if this configuration hadn't applied yet.
+
+## Design
+
+### Request/response
+The extension uses request/response pattern. A response from client is sent to confirm delivery.
+
+For MQTT, responses MUST be published at `<request_path>/status`. Each response MUST be published with the same QoS as the corresponding request.
+
+### Formats
+#### Schemeless JSON
+##### Use case 1
+The server should be able to listen for configuration request at the following resource path:
+```
+<endpoint_token>/pull/json
+```
+
+The payload should be a JSON-encoded object with the following fields:
+- `id` (required) - id of the message. Should be either string or number. Used in delivery confirmation process.
+- `configVersion` - current endpoint configuration version. Should be of integer type. Used by server to determine is it needed to send response with configuration, server must not send configuration record if latest available configuration version equals to the one sent by endpoint. If this field was excluded from the message, then server will send configuration anyway.
+- `config` - configuration body of an arbitrary JSON type.
+
+Example:
+```json
+{
+  "id": 42,
+  "configVersion": 1,
+  "config": {
+      "key": "value",
+      "array" : [
+          "value2"
+      ]
+  }
+}
+```
+
+A server response is a JSON record with the following fields:
+- `id` a copy of the `id` field from the corresponding request.
+- `configVersion` (required) - version of configuration that is included into the message. If there's no new configuration versions and config body isn't included into message, then value of this field will equal to the one provided by endpoint.
+- `status` a human-readable string explaining the cause of an error (if any). In case that request was sucessful, it is `"ok"`.
+- `config` - configuration body of an arbitrary JSON type.
+Destination topic is 
+```
+<endpoint_token>/push/json
```

Comment:
Seems like I was not clear enough. My comment was for this particular use case when the configuration is requested by endpoint.

I see two scenarious:
1. Endpoint requests configuration from server.
In this case Endpoint makes /pull, server responds with /status

2. Server pushes configuration to endpoint.
Here, Server makes /push, Endpoint responds with /status.

Such schemas will perfectly match the request/response definitions above and at the same time will be clear what configuration delivery scenario is used.